README.md
# shorty

[![Latest version](https://img.shields.io/crates/v/shorty.svg)](https://crates.io/crates/shorty)
[![Build Status](https://travis-ci.org/ffissore/shorty.svg?branch=master)](https://travis-ci.org/ffissore/shorty)
![License](https://img.shields.io/github/license/ffissore/shorty.svg)

Shorty is a URL shortener: it assigns a short ID to a URL of any length, and when people will access the URL with that short ID, they will be redirected to the original URL.

This is useful in cases such as sending SMS notifications, when you have a limited number of characters and don't want to waste them with a long URL and its parameters.

Shorty stores its data on Redis.

### Multiple ways of deploying it

Shorty is available as

- a [rust library](#rust-library)
- an [http microservice](#http-microservice)
- an [AWS lambda](#aws-lambda)
- an [Azure function](#azure-function)

### Rust library

Shorty is written in rust, and available as a crate library.

For additional information, take a look at the [documentation](https://docs.rs/shorty), and use shorty-http binary crate as an example. 

### HTTP microservice

Shorty stores its data on redis, so you need to install redis first. How to do that depends on your operating system. If you are on a debian like linux distro, it's just a

```bash
apt-get install redis
```

If you prefer `docker`, it's

```bash
docker run --rm -it -p 6379:6379 redis:alpine
```

Once redis is ready, download the latest shorty [release](https://github.com/ffissore/shorty/releases), unpack the archive and run shorty with

```bash
SHORTENER_API_KEY_MANDATORY=false ./shorty-http
```

Again, if you prefer `docker`, it's

```bash
docker run --rm -it -e SHORTENER_API_KEY_MANDATORY=false --network host ffissore/shorty
```

Shorty will log `Starting server on 127.0.0.1:8088`.

Now scroll down to [Using shorty](#using-shorty).

### AWS lambda

In order to deploy Shorty on AWS, you need... node.js. Duh! Yeah, it's a shame, but `serverless` is a node package and de-facto standard for deploying lambdas on AWS, and it works well.

Install the required dependencies with 

```bash
npm install
```

Then run

```bash
npx serverless deploy
```

This will download a docker image that will be used to compile shorty. It will then create a Redis instance on AWS, all the networking bits required to make the lambda connect to Redis, and finally it will deploy the lambda.

When done, it will print something like

```
endpoints:
  GET - https://9tgwceucu4.execute-api.us-east-1.amazonaws.com/dev/{key}
  POST - https://9tgwceucu4.execute-api.us-east-1.amazonaws.com/dev/
```

Please bear in mind that:
1. Your AWS account will be charged: lambdas are free up to 1 million requests but the "NAT gateway" required to expose them is not
1. The bloat of CloudFormation you'll find in `serverless.yml` has been copy-pasted (thx [ittus](https://github.com/ittus/aws-lambda-vpc-nat-examples/blob/master/serverless.yml)) and I barely understand what it does.

When you're done playing with shorty, delete everything with
```bash
npx serverless remove
```   

Now scroll down to [Using shorty](#using-shorty).

### Azure function

Azure does not support Rust directly, but they allow you to run any kind of runtime as long as it runs in docker.

Shorty docker image is available at [docker hub](https://hub.docker.com/r/ffissore/shorty).

The following commands were taken from [this Azure guide](https://docs.microsoft.com/azure/app-service/containers/tutorial-custom-docker-image).

* Create a resource group
```bash
az group create --name shorty-resources --location "West Europe"
```
* Create a redis instance  
```bash
az redis create --resource-group shorty-resources --name shorty-redis --location "West Europe" --sku Basic --vm-size c0 --enable-non-ssl-port
```
* Create a service plan
```bash
az appservice plan create --name shorty-service-plan --resource-group shorty-resources --sku B1 --is-linux
``` 
* Create a webapp
```bash
az webapp create --resource-group shorty-resources --plan shorty-service-plan --name shortyshorty --deployment-container-image-name ffissore/shorty:latest
```
* Tell azure that Shorty listens on port 8088
```bash
az webapp config appsettings set --resource-group shorty-resources --name shortyshorty --settings WEBSITES_PORT=8088
```
* and that for now we don't need API keys
```bash
az webapp config appsettings set --resource-group shorty-resources --name shortyshorty --settings SHORTENER_API_KEY_MANDATORY=false
```
* Now locate your redis instance on azure portal, and copy its primary access key. It will something like `RandomString=`: that's redis password. Set the redis host variable accordingly
```bash
az webapp config appsettings set --resource-group shorty-resources --name shortyshorty --settings SHORTENER_REDIS_HOST=:RandomString=@shorty-redis.redis.cache.windows.net
```

Shorty will be available at https://shortyshorty.azurewebsites.net/

Please bear in mind that your account will be charged for all of the above.

When you're done playing with shorty, delete everything with
```bash
az group delete --name shorty-resources
```   

Now scroll down to [Using shorty](#using-shorty).
 
### Using shorty

The following instructions assume shorty is running on your pc. If that's not the case, replace `http://localhost:8088` with the proper domain.
 
Try this `curl` to store a URL

```bash
curl -vv http://localhost:8088/ -H 'Content-Type: application/json' --data '{"url":"https://en.wikipedia.org/wiki/URL_shortening#Techniques"}'
```

If you have enabled API keys, add the API key to the payload, like this

```bash
curl -vv http://localhost:8088/ -H 'Content-Type: application/json' --data '{"api_key": "test", "url":"https://en.wikipedia.org/wiki/URL_shortening#Techniques"}'
```

It will output something like

```json
{"id":"CGQ6LM8bfj","url":"https://en.wikipedia.org/wiki/URL_shortening#Techniques"}
```

Now try resolving that ID

```bash
curl -vv http://localhost:8088/CGQ6LM8bfj
```

The output headers of curl will contain a `Location: https://en.wikipedia.org/wiki/URL_shortening#Techniques`. Try opening the shorty url with your browser.

### Configuration

Shorty can be configured through environment variables

* `SHORTENER_REDIS_HOST`: the host of the redis server, defaults to 127.0.0.1
* `SHORTENER_REDIS_PORT`: the port of the redis server, defaults to 6379
* `SHORTENER_API_KEY_MANDATORY`: do users have to provide an API key in order to create a new short URL? boolean, defaults to true
* `SHORTENER_RATE_LIMIT`: the amount of new short url a single API key can create in a period, defaults to 10, if set to 0 no limit is applied
* `SHORTENER_RATE_LIMIT_PERIOD`: the period of the rate limit, if active, defaults to 600 seconds (10 mins)
* `SHORTENER_ID_LENGTH`: the length of the ID generated for each URL, defaults to 10. The char set is `a-zA-Z0-9` = 62 chars. If you plan to use shorty only internally, you can use a much shorter ID, like 4 chars.
* `SHORTENER_ID_GENERATION_MAX_ATTEMPTS`: the max number of attempts to generate a unique ID, defaults to 10. Especially important when the ID length is short and many short URLs are created.
* `SHORTENER_HOST`: the host shorty will listen to
* `SHORTENER_PORT`: the port shorty will listen to

### What's on Redis

* API keys: they are prefixed with `API_KEY_`, stored as `API_KEY_my_api_key`, and assigned a boolean value. A missing API key or an API key assigned to `false` will return error "Invalid API key"
* Call rate keys: they are prefixed with `RATE_`, stored as `RATE_my_api_key`, and assigned the registered number of calls. The key is valid until `rate limit period` (see paragraph above) is over.
* Short IDs, at the configured length (see example above): they are assigned to the original URL


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.5.4] - 2020-06-15
### Changed
- Removed links to with.lv: I took it down because it was abused
- Updated dependencies

## [0.5.3] - 2020-01-24
### Security
- Security fixes

## [0.5.2] - 2019-09-03
### Security
- Security fixes

## [0.5.1] - 2019-07-09
### Security
- Security fixes

## [0.5.0] - 2019-04-18
### Added
- Parsing URLs before shortening them, therefore checking that they are valid
- Link loop check: an attempt to shorten a url with the same domain used by shorty to serve shortened urls will result in an error

## [0.4.0] - 2019-04-10
### Fixed
- Another `RedisFacade.get_bool` bug
### Added
- Conflicting ID check and multiple attempts to generate a unique one

## [0.3.0] - 2019-04-09
### Fixed
- `RedisFacade.get_bool` bug
### Added
- Documentation

## [0.2.1] - 2019-04-09
### Added
- Also publishing docker `latest` image

## [0.2.0] - 2019-04-09
- First implementation


shorty/src/lib.rs
// Copyright 2019 Federico Fissore
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! # Shorty
//!
//! `shorty` is a URL shortener: it assigns a short ID to a URL of any length, and when people will
//! access the URL with that short ID, they will be redirected to the original URL.
//!
#[macro_use]
extern crate serde_derive;

use core::fmt;
use std::error::Error;
use std::fmt::{Display, Formatter};

use redis::{ErrorKind, RedisError};
use url::Url;

#[cfg(test)]
use tests::StubRedisFacade as RedisFacade;

#[cfg(not(test))]
use crate::redis_facade::RedisFacade;

#[cfg(not(test))]
pub mod redis_facade;

#[derive(Debug)]
pub struct ShortenerError {
    message: &'static str,
    cause: Option<Box<dyn Error>>,
}

impl ShortenerError {
    fn new(message: &'static str) -> ShortenerError {
        ShortenerError {
            message,
            cause: None,
        }
    }
    fn new_with_cause(message: &'static str, error: Box<dyn Error>) -> ShortenerError {
        ShortenerError {
            message,
            cause: Some(error),
        }
    }
}

impl Display for ShortenerError {
    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
        f.write_str(&self.message)?;

        if let Some(err) = &self.cause {
            return f.write_str(&format!(" - {}", &err));
        }

        Ok(())
    }
}

impl Error for ShortenerError {}

/// `Shortener` is the struct exposing methods `lookup` and `shorten`.
///
/// `lookup` attempts to resolve an ID to a URL. If no URL is found or an error occurs, it returns
/// `None`, otherwise it returns `Some(url)`.
///
/// `shorten` takes an optional API key and a URL to shorten. If the API key is present, it will
/// validate it and shorten the URL only if validation passes. Otherwise, it will just shorten the
/// URL.
///
/// `Shortener` interacts with a `RedisFacade`, which makes it easier to work with the `redis` crate
/// and simplifies testing.
pub struct Shortener {
    id_length: usize,
    id_alphabet: Vec<char>,
    id_generation_max_attempts: u8,
    redis: RedisFacade,
    rate_limit_period: usize,
    rate_limit: i64,
}

/// A struct with the successful result of a URL shortening. It holds the original `url` and the
/// resulting `id`
#[derive(Serialize)]
pub struct ShortenerResult {
    id: String,
    url: String,
}

impl Shortener {
    /// Creates a new Shortener
    ///
    /// `id_length` is the length of the generated ID.
    ///
    /// `id_alphabet` is the alphabet used in the ID: a decent one is `a-zA-Z0-9` as each entry has
    /// 62 possible values and is ASCII
    ///
    /// `id_generation_max_attempts` is the number of attempts to generate an unique ID when a
    /// conflict is detected.
    ///
    /// `redis` is a `RedisFacade` instance.
    ///
    /// `rate_limit_period` is the amount of seconds during which calls to `shorten` will be counted.
    ///
    /// `rate_limit` is the max number of calls that can be made to `shorten` in a period.
    pub fn new(
        id_length: usize,
        id_alphabet: Vec<char>,
        id_generation_max_attempts: u8,
        redis: RedisFacade,
        rate_limit_period: usize,
        rate_limit: i64,
    ) -> Shortener {
        Shortener {
            id_length,
            id_alphabet,
            id_generation_max_attempts,
            redis,
            rate_limit_period,
            rate_limit,
        }
    }

    /// Looks up a URL by the given ID. If no URL is found or an error occurs, it returns `None`,
    /// otherwise it returns `Some(url)`.
    pub fn lookup(&self, id: &str) -> Option<String> {
        match self.redis.get_string(id) {
            Ok(url) => Some(url),
            Err(_) => None,
        }
    }

    fn verify_api_key(&self, api_key: &str) -> Result<(), ShortenerError> {
        let api_key = format!("API_KEY_{}", api_key);
        log::trace!("verifying api key '{}'", api_key);

        let verify_and_increment = self.redis.get_bool(&api_key).and_then(|valid| {
            if !valid {
                return Err(RedisError::from((
                    ErrorKind::ExtensionError,
                    "API key expired",
                )));
            }

            if self.rate_limit <= 0 {
                return Ok(-1);
            }

            let rate_key = format!("RATE_{}", api_key);
            log::trace!("verifying rate key '{}'", rate_key);

            self.redis.exists(&rate_key).and_then(|exists| {
                log::trace!("rate key exists {}", exists);

                self.redis.increment(&rate_key).and_then(|number_of_calls| {
                    log::trace!("rate key {} number of calls {}", rate_key, number_of_calls);

                    if !exists {
                        self.redis
                            .expire(&rate_key, self.rate_limit_period)
                            .unwrap();
                    }

                    Ok(number_of_calls)
                })
            })
        });

        match verify_and_increment {
            Ok(call_rate) if self.rate_limit > 0 && call_rate > self.rate_limit => {
                Err(ShortenerError::new("Rate limit exceeded"))
            }
            Ok(_) => Ok(()),
            Err(err) => Err(ShortenerError::new_with_cause(
                "Invalid API key",
                Box::new(err),
            )),
        }
    }

    fn generate_id(&self) -> Result<String, ShortenerError> {
        for _ in 1..=self.id_generation_max_attempts {
            let id = nanoid::custom(self.id_length, &self.id_alphabet);

            let exists = self.redis.exists(&id).unwrap_or(false);

            if !exists {
                return Ok(id);
            }
        }

        Err(ShortenerError::new(
            "Failed to generate an ID: too many attempts. Consider using a longer ID",
        ))
    }

    /// Shortens an URL, returning a `ShortenerResult` holding the provided URL and the generated ID.
    ///
    /// If the optional API key is present, it will validate it and shorten the URL only if
    /// validation passes.
    ///
    /// If the optional host is present, it will ensure that the url to shorten is not a url from
    /// the same host that's running shorty (which would create a link loop)
    ///
    /// Otherwise, it will just shorten the URL.
    pub fn shorten(
        &self,
        api_key: &Option<&str>,
        host: Option<&str>,
        url: &str,
    ) -> Result<ShortenerResult, ShortenerError> {
        let verify_result = api_key
            .as_ref()
            .map(|api_key| self.verify_api_key(api_key))
            .unwrap_or(Ok(()));

        verify_result
            .and_then(|_| self.generate_id())
            .and_then(|id| {
                let mut url = url.to_owned();
                if !url.to_lowercase().starts_with("http") {
                    url = format!("http://{}", url);
                }
                Url::parse(&url)
                    .and_then(|parsed_url| Ok((id, url, parsed_url)))
                    .map_err(|parse_err| {
                        ShortenerError::new_with_cause("Unable to parse url", Box::new(parse_err))
                    })
            })
            .and_then(|(id, url, parsed_url)| {
                if host.is_none() {
                    return Ok((id, url));
                }

                if parsed_url.host_str().unwrap().eq(host.unwrap()) {
                    return Err(ShortenerError::new("Link loop is not allowed"));
                }

                Ok((id, url))
            })
            .and_then(|(id, url)| {
                self.redis
                    .set(&id, url.as_str())
                    .map(|_| ShortenerResult { id, url })
                    .map_err(|err| ShortenerError::new_with_cause("Redis error", Box::new(err)))
            })
    }
}

#[cfg(test)]
mod tests {
    use std::cell::RefCell;

    use redis::RedisResult;

    use super::*;

    pub struct StubRedisFacade {
        get_string_answers: RefCell<Vec<RedisResult<String>>>,
        get_bool_answers: RefCell<Vec<RedisResult<bool>>>,
        exists_answers: RefCell<Vec<RedisResult<bool>>>,
        set_answers: RefCell<Vec<RedisResult<()>>>,
        incr_answers: RefCell<Vec<RedisResult<i64>>>,
        expire_answers: RefCell<Vec<RedisResult<()>>>,
    }

    impl StubRedisFacade {
        fn new() -> Self {
            StubRedisFacade {
                get_string_answers: RefCell::new(vec![]),
                get_bool_answers: RefCell::new(vec![]),
                exists_answers: RefCell::new(vec![]),
                set_answers: RefCell::new(vec![]),
                incr_answers: RefCell::new(vec![]),
                expire_answers: RefCell::new(vec![]),
            }
        }

        pub fn get_string(&self, _key: &str) -> RedisResult<String> {
            if self.get_string_answers.borrow().len() > 0 {
                return self.get_string_answers.borrow_mut().remove(0);
            }
            panic!("unexpected get_string call");
        }

        pub fn get_bool(&self, _key: &str) -> RedisResult<bool> {
            if self.get_bool_answers.borrow().len() > 0 {
                return self.get_bool_answers.borrow_mut().remove(0);
            }
            panic!("unexpected get_bool call");
        }

        pub fn exists(&self, _key: &str) -> RedisResult<bool> {
            if self.exists_answers.borrow().len() > 0 {
                return self.exists_answers.borrow_mut().remove(0);
            }
            panic!("unexpected exists call");
        }

        pub fn set(&self, _key: &str, _value: &str) -> RedisResult<()> {
            if self.set_answers.borrow().len() > 0 {
                return self.set_answers.borrow_mut().remove(0);
            }
            panic!("unexpected set call");
        }

        pub fn increment(&self, _key: &str) -> RedisResult<i64> {
            if self.incr_answers.borrow().len() > 0 {
                return self.incr_answers.borrow_mut().remove(0);
            }
            panic!("unexpected increment call");
        }

        pub fn expire(&self, _key: &str, _seconds: usize) -> RedisResult<()> {
            if self.expire_answers.borrow().len() > 0 {
                return self.expire_answers.borrow_mut().remove(0);
            }
            panic!("unexpected expire call");
        }
    }

    #[test]
    fn test_lookup() {
        let redis = StubRedisFacade::new();
        &redis
            .get_string_answers
            .borrow_mut()
            .push(Ok(String::from("test url")));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, 10);
        assert_eq!(shortener.lookup("id").unwrap(), "test url");
    }

    #[test]
    fn test_shorten_happy_path_first_call() {
        let redis = StubRedisFacade::new();
        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(true));
        &redis.exists_answers.borrow_mut().push(Ok(false));
        &redis.incr_answers.borrow_mut().push(Ok(1));
        &redis.expire_answers.borrow_mut().push(Ok(()));

        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        // shortened url storage
        &redis.set_answers.borrow_mut().push(Ok(()));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, 10);
        let shorten_result = shortener
            .shorten(&Some("api key"), Some("with.lv"), "example.com")
            .unwrap();
        assert_eq!(10, shorten_result.id.len());
        assert_eq!("http://example.com", shorten_result.url);
    }

    #[test]
    fn test_shorten_happy_path_no_rate_limit() {
        let redis = StubRedisFacade::new();
        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(true));

        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        // shortened url storage
        &redis.set_answers.borrow_mut().push(Ok(()));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, -1);
        let shorten_result = shortener
            .shorten(&Some("api key"), Some("with.lv"), "example.com")
            .unwrap();
        assert_eq!(10, shorten_result.id.len());
        assert_eq!("http://example.com", shorten_result.url);
    }

    #[test]
    fn test_shorten_happy_path_second_call() {
        let redis = StubRedisFacade::new();
        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(true));
        &redis.exists_answers.borrow_mut().push(Ok(true));
        &redis.incr_answers.borrow_mut().push(Ok(2));

        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        // shortened url storage
        &redis.set_answers.borrow_mut().push(Ok(()));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, 10);
        let shorten_result = shortener
            .shorten(&Some("api key"), Some("with.lv"), "example.com")
            .unwrap();
        assert_eq!(10, shorten_result.id.len());
        assert_eq!("http://example.com", shorten_result.url);
    }

    #[test]
    fn test_shorten_happy_path_no_api_key() {
        let redis = StubRedisFacade::new();
        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        // shortened url storage
        &redis.set_answers.borrow_mut().push(Ok(()));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, 10);
        let shorten_result = shortener
            .shorten(&None, Some("with.lv"), "example.com")
            .unwrap();
        assert_eq!(10, shorten_result.id.len());
        assert_eq!("http://example.com", shorten_result.url);
    }

    #[test]
    fn test_shorten_unhappy_path_rate_limit_exceeded() {
        let rate_limit = 10;
        let redis = StubRedisFacade::new();
        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(true));
        &redis.exists_answers.borrow_mut().push(Ok(true));
        &redis.incr_answers.borrow_mut().push(Ok(rate_limit + 1));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, rate_limit);
        let shorten_result_err = shortener
            .shorten(&Some("api key"), Some("with.lv"), "example.com")
            .err()
            .unwrap();
        assert_eq!("Rate limit exceeded", shorten_result_err.message);
    }

    #[test]
    fn test_shorten_unhappy_path_bad_url() {
        let redis = StubRedisFacade::new();
        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, -1);
        let shorten_result_err = shortener
            .shorten(&None, Some("with.lv"), "wrong domain.com")
            .err()
            .unwrap();
        assert_eq!("Unable to parse url", shorten_result_err.message);
    }

    #[test]
    fn test_shorten_unhappy_path_same_domain() {
        let redis = StubRedisFacade::new();
        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, -1);
        let shorten_result_err = shortener
            .shorten(&None, Some("example.com"), "example.com")
            .err()
            .unwrap();
        assert_eq!("Link loop is not allowed", shorten_result_err.message);
    }

    #[test]
    fn test_shorten_happy_path_rate_limit_expired() {
        let redis = StubRedisFacade::new();

        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(true));
        &redis.exists_answers.borrow_mut().push(Ok(true));
        &redis.incr_answers.borrow_mut().push(Ok(1));

        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        // shortened url storage
        &redis.set_answers.borrow_mut().push(Ok(()));

        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(true));
        &redis.exists_answers.borrow_mut().push(Ok(false));
        &redis.incr_answers.borrow_mut().push(Ok(1));
        &redis.expire_answers.borrow_mut().push(Ok(()));

        // id generation
        &redis.exists_answers.borrow_mut().push(Ok(false));

        // shortened url storage
        &redis.set_answers.borrow_mut().push(Ok(()));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, 10);

        let shorten_result = shortener
            .shorten(&Some("api key"), Some("with.lv"), "example.com")
            .unwrap();
        assert_eq!(10, shorten_result.id.len());
        assert_eq!("http://example.com", shorten_result.url);

        let shorten_result = shortener
            .shorten(&Some("api key"), Some("with.lv"), "www.wikipedia.org")
            .unwrap();
        assert_eq!(10, shorten_result.id.len());
        assert_eq!("http://www.wikipedia.org", shorten_result.url);
    }

    #[test]
    fn test_shorten_unhappy_path_invalid_api_key() {
        let redis = StubRedisFacade::new();

        // api key verification
        &redis.get_bool_answers.borrow_mut().push(Ok(false));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 10, redis, 600, 10);
        let shorten_result_err = shortener
            .shorten(&Some("api key"), Some("with.lv"), "example.com")
            .err()
            .unwrap();
        assert_eq!("Invalid API key", shorten_result_err.message);
    }

    #[test]
    fn test_shorten_unhappy_path_too_many_attempts_generating_id() {
        let redis = StubRedisFacade::new();

        // id generation attempts
        &redis.exists_answers.borrow_mut().push(Ok(true));
        &redis.exists_answers.borrow_mut().push(Ok(true));

        let shortener = Shortener::new(10, vec!['a', 'b', 'c'], 2, redis, 600, 10);
        let shorten_result_err = shortener
            .shorten(&None, Some("with.lv"), "example.com")
            .err()
            .unwrap();
        assert_eq!(
            "Failed to generate an ID: too many attempts. Consider using a longer ID",
            shorten_result_err.message
        );
    }
}
